### DISCUSSION TITLE
Can a plugin supply arbitrary functions that can be called?

### DISCUSSION THREAD

**Discussion by BanksySan** [artilleryio/artillery#1934] [Artillery]
When I have a processor, I can call some function in from the script.

Can a plugin provide the same sort of behavior?

I'd like the script to be able to call a function like `logIn` or `logOut` that I define in the plugin using the `function` instruction.

**Comment by BanksySan** [comment]
I can see that the processor is loaded in the [runner.js](https://github.com/artilleryio/artillery/blob/c2a39fe4be80707eff162f14b9759cc21014ccfc/packages/core/lib/runner.js#LL80C1-L80C23).  However this happens after the plugin is loaded, so I can't add functions to it and it will wipe out any functions I add in the plugin.

**Comment by hassy** [comment]
hi @BanksySan! Yep, there's a way to do it.

Plugins have full access to the entire test script. So inside a plugin, you can attach arbitrary functions to the `processor` property, and those functions become available to scenarios.

There's an example in the built-in `expectations` plugin:

https://github.com/artilleryio/artillery/blob/main/packages/artillery-plugin-expect/index.js#L33-L35

Here the plugin makes sure that `script.config.processor` can be extended.

And then it attaches a function to it:

https://github.com/artilleryio/artillery/blob/main/packages/artillery-plugin-expect/index.js#L50

That means that scenarios could use the "expectationsPluginCheckExpectations" function now.

That's not the intended usage of the `expect` plugin, but the same mechanism should work for your use case.

**Reply by BanksySan** [reply]
Thanks.  I've just tried it and it looks good.
